File: day-23/day-23.r
# advent of code
# day 23

# part 1
input <- '135468729'

predictCups <- function(input, k = 100, l = 9){
  sequence <- 1:l
  header <- as.integer(unlist(strsplit(input, '')))
  sequence[1:length(header)] <- header
  max <- max(sequence)
  for(i in 1:k){
    if(i %% 1000 == 0){print(i)}
    current <- sequence[1]
    removed <- sequence[2:4]
    if(current == 1){
      target <- max
    } else {
      target <- current - 1
    }
    while(target %in% removed){
      if(target == 1){
        target <- max
      } else {
        target <- target - 1
      }
    }
    if(which(sequence == target) < length(sequence)){
      post <- sequence[(which(sequence == target) + 1):length(sequence)]
    } else {
      post <- integer()
    }
    sequence <- c(sequence[5:which(sequence == target)], removed, post, current)
  }
  if(k == 100 & l == 9){
    if(which(sequence == 1) == 1){
      post <- integer()
    } else {
      post <- sequence[1:(which(sequence == 1) - 1)]
    }
    if(which(sequence == 1) == length(sequence)){
      pre <- integer()
    } else {
      pre <- sequence[(which(sequence == 1) + 1):length(sequence)]
    }
    return(paste(c(pre, post), collapse = ''))
  } else {
    if(which(sequence == 1) == 1){
      post <- integer()
    } else {
      post <- sequence[1:(which(sequence == 1) - 1)]
    }
    if(which(sequence == 1) == length(sequence)){
      pre <- integer()
    } else {
      pre <- sequence[(which(sequence == 1) + 1):length(sequence)]
    }
    sequence <- c(pre, post)
    return(as.character(prod(sequence[1:2])))
  }
}

predictCups(input)

# part 2

predictCups(input, k = 10000000, l = 1000000)

File: day-23/day-23.py
# advent of code
# day 23

# part 1
import numpy as np

input = '135468729'

def predictCups(input, partTwo = False):
    if partTwo == True:
        l = 1000000
        k = 10000000
    else:
        l = 9
        k = 100
    sequence = list(range(1, l + 1))
    sequence[:9] = [int(char) for char in input]
    max = l
    for i in range(k):
        if(i % 1000 == 0):
            print(i)
        current = sequence[0]
        removed = sequence[1:4]
        for p in range(3):
            sequence.pop(1)
        if(current == 1):
            target = max
        else:
            target = current - 1
        while(target in removed):
            if(target == 1):
                target = max
            else:
                target -= 1
        for x in removed[::-1]:
            if(sequence.index(target) == len(sequence) - 1):
                sequence.append(x)
            else:
                sequence.insert(sequence.index(target) + 1, x)
        sequence.pop(0)
        sequence.append(current)
    oneLoc = sequence.index(1)
    sequence = sequence[oneLoc:] + sequence[:oneLoc]
    sequence.pop(0)
    if partTwo == True:
        return(np.prod(sequence[:2]))
    else:
        strAnswer = ''.join([str(x) for x in sequence])
        return(strAnswer)

print(predictCups(input, partTwo = False))

# part 2
predictCups(input, partTwo = True)
